### PR TITLE
deploy and test files added

### PR DESCRIPTION
**Pull request by Shivansh070707**

### CODE REVIEW

**Review by sumit03guha — CHANGES_REQUESTED** [review]
### Test cases
- Add more test cases to assert the storage variables using the `StratX2Getter` contract.
- Rename the `StratX2Getter.Facet` to `StratX2GetterFacet`.

### Contracts
- It is better to append the term `Facet` after every facet contract.
- It is better to create interfaces separately in the `interfaces` folder. https://github.com/Shivansh070707/Tasks/blob/cc465b63d82604805c56fb427b791db2244dddce/Task4/contracts/diamond/facets/StratX2Facet.sol#L11
- Use functions instead of modifiers. It helps in saving gas. https://github.com/Shivansh070707/Tasks/blob/cc465b63d82604805c56fb427b791db2244dddce/Task4/contracts/diamond/facets/StratX2Facet.sol#L19
- Remove the  console logs after testing. https://github.com/Shivansh070707/Tasks/blob/cc465b63d82604805c56fb427b791db2244dddce/Task4/contracts/diamond/facets/StratX2Facet.sol#L9
- Instead of making the `StratX2Facet` as pausable, create a Pausable storage struct in the diamond and use it in the facets.
- Use different revert strings within the same function. https://github.com/Shivansh070707/Tasks/blob/cc465b63d82604805c56fb427b791db2244dddce/Task4/contracts/diamond/facets/StratX2Facet.sol#L324 https://github.com/Shivansh070707/Tasks/blob/cc465b63d82604805c56fb427b791db2244dddce/Task4/contracts/diamond/facets/StratX2Facet.sol#L325

### Deployment
- Add functionality to save the addresses and ABIs of the deployments.

File: Task4/scripts/deploy.ts [hunk 1]
```diff
@@ -59,49 +144,97 @@ export async function deployDiamond() {
     });
   }
 
-  // console.log("Diamond Cut: ", cut);
+  //console.log('Diamond Cut: ', cut);
   const diamondCut = await ethers.getContractAt('IDiamondCut', diamond.address);
   const functionCall = diamondInit.interface.encodeFunctionData('init', [
     [
       '0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c',
```
Comment (sumit03guha) [inline Task4/scripts/deploy.ts:151]: use descriptive names for the addresses by declaring a named variable

File: Task4/scripts/deploy.ts [hunk 2]
```diff
@@ -59,49 +144,97 @@ export async function deployDiamond() {
...
-      '0x000000000000000000000000000000000000dEaD',
+      owner.address,
+      farmA.address,
+      autoV2.address,
+      want.address,
+      matic.address,
+      bitcoin.address,
+      autoV21.address,
+      farmB.address,
       '0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D',
```
Comment (sumit03guha) [inline Task4/scripts/deploy.ts:160]: use descriptive names for the addresses by declaring a named variable

File: Task4/scripts/deploy.ts [hunk 3]
```diff
@@ -59,49 +144,97 @@ export async function deployDiamond() {
...
+    'Build/StratX2GetterFacet.json',
+    JSON.stringify(StratX2GetterData, null, '\t')
+  );
 
   console.log('**** Diamond deploy end');
-  return diamond.address;
+  return {
+    matic,
+    bitcoin,
+
```
Comment (sumit03guha) [inline Task4/scripts/deploy.ts:224]: remove white spaces

File: Task4/contracts/diamond/libraries/LibDiamond.sol [hunk 4]
```diff
@@ -31,9 +31,11 @@ library LibDiamond {
         address contractOwner;
     }
     struct StratX2Storage {
+        bool _paused;
         bool isCAKEStaking; // only for staking CAKE using pancakeswap's native CAKE staking contract.
         bool isSameAssetDeposit;
-        bool isAutoComp; // this vault is purely for staking. eg. WBNB-AUTO staking vault.
+        bool isAutoComp;
+        uint8 lock; // this vault is purely for staking. eg. WBNB-AUTO staking vault.
```
Comment (sumit03guha) [inline Task4/contracts/diamond/libraries/LibDiamond.sol:38]: if `lock` is not used anywhere, remove it

File: Task4/scripts/deploy.ts [hunk 5]
```diff
@@ -17,37 +43,96 @@ export async function deployDiamond() {
...
+    address: diamondCutFacet.address,
+    network: {
+      name: diamondCutFacet.provider._network.name,
+      chainId: diamondCutFacet.provider.network.chainId,
+    },
+    abi: JSON.parse(diamondCutFacet.interface.format('json')),
+  };
+  fs.writeFileSync(
+    'Build/DiamondCutFacet.json',
+    JSON.stringify(diamondCutFacetData, null, '\t')
```
Comment (sumit03guha) [inline Task4/scripts/deploy.ts:69]: instead of `\t` use `2` indentation while saving JSON

File: Task4/hardhat.config.ts [hunk 6]
```diff
@@ -48,7 +43,7 @@ const config: HardhatUserConfig = {
   contractSizer: {
     alphaSort: true,
     disambiguatePaths: false,
-    runOnCompile: false,
+    runOnCompile: true,
```
Comment (sumit03guha) [inline Task4/hardhat.config.ts:46]: better to keep the `runOnCompile` as `false`.

**Review by sumit03guha — CHANGES_REQUESTED** [review]
1. Use better filenames instead of `index.ts`. Example : `facet.test.ts`
2. Remove unnecessary white spaces.

File: Task4/scripts/deploy.ts (hunk 1 above)
Comment (Shivansh070707) [reply Task4/scripts/deploy.ts]: already up

**Pull request merged**